### PR TITLE
Cbor plugin system

### PR DESCRIPTION
**Pull request by david415**
Can we have a CBOR based plugin system for the Katzenpost mix server? Isn't CBOR way less annoying than gRPC.

### CODE REVIEW

File: cborplugin/client.go [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+	"github.com/katzenpost/core/worker"
+	"github.com/ugorji/go/codec"
+	"gopkg.in/op/go-logging.v1"
+)
+
+// KaetzchenPluginInterface is the interface that we expose for external
+// plugins to implement. This is similar to the internal Kaetzchen
+// interface defined in:
+// github.com/katzenpost/server/internal/provider/kaetzchen/kaetzchen.go
+type KaetzchenPluginInterface interface {
```
Comment (willscott) [inline cborplugin/client.go]: `Interface` in the interface name is a bit redundant.

File: cborplugin/client.go [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+	OnRequest(id uint64, request []byte, hasSURB bool) ([]byte, error)
+
+	// Parameters returns the agent's paramenters for publication in
+	// the Provider's descriptor.
+	Parameters() map[string]string
+
+	Halt()
+}
+
+type Request struct {
```
Comment (willscott) [inline cborplugin/client.go]: should `OnRequest` above directly take the `Request` struct for consistency?

File: cborplugin/client.go [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+	}
+	c.Go(c.worker)
+	return nil
+}
+
+func (c *Client) worker() {
+	<-c.HaltCh()
+	c.cmd.Process.Signal(syscall.SIGHUP)
+	err := c.cmd.Wait()
+	if err != nil {
```
Comment (willscott) [inline cborplugin/client.go]: log or return an error

File: cborplugin/client.go [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+		// XXX log err here?
+	}
+}
+
+func (c *Client) dial(socketPath string) {
```
Comment (willscott) [inline cborplugin/client.go]: should this not be part of the public interface? (capitalized)

File: internal/provider/kaetzchen/cbor_plugins.go [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+
+func (k *CBORPluginWorker) launch(command string, args []string) (*cborplugin.Client, error) {
+	k.log.Debugf("Launching plugin: %s", command)
+	plugin := cborplugin.New(k.log)
+	err := plugin.Start(command, args)
+	return plugin, err
+}
+
+// NewCBORPluginWorker returns a new CBORPluginWorker
+func NewCBORPluginWorker(glue glue.Glue) (*CBORPluginWorker, error) {
```
Comment (willscott) [inline internal/provider/kaetzchen/cbor_plugins.go]: glue isn't a great name for the object. can this be a static method?

File: cborplugin/client.go [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+// github.com/katzenpost/server/internal/provider/kaetzchen/kaetzchen.go
+type KaetzchenPluginInterface interface {
+	// OnRequest is the method that is called when the Provider receives
+	// a request desgined for a particular agent. The caller will handle
+	// extracting the payload component of the message
+	OnRequest(id uint64, request []byte, hasSURB bool) ([]byte, error)
+
+	// Parameters returns the agent's paramenters for publication in
+	// the Provider's descriptor.
+	Parameters() map[string]string
```
Comment (willscott) [inline cborplugin/client.go]: These end up published in `forPKI`, but the purpose here isn't totally clear. more structure would be useful, probably. Are there maximum expected lengths of key-values? could this be fully opaque, and if not, what other part of the system is expecting to understand these semantics besides the plugin?

Comment (david415) [reply cborplugin/client.go]: Actually, I think we need to improve our documentation of this feature a bit. It is mentioned here in the spec:
https://github.com/katzenpost/docs/blob/master/specs/kaetzchen.rst#4-pki-extensions

...but I think that spec should specificy the method names of the Kaetzchen interface.
Also I think the name Kaetzchen is confusing so I've just been calling them "services" or sometimes "micro-services" if I want to see what it feels like to say a popular buzz word.

Basically the output of Parameters() ends up being published in a map of associated with the service name. This map is part of the Mix Descriptor which is here:
https://github.com/katzenpost/core/blob/master/pki/pki.go#L247

File: internal/provider/kaetzchen/cbor_plugins.go (hunk 5 above)
Comment (david415) [reply internal/provider/kaetzchen/cbor_plugins.go]: yeah glue is a horrible name but i'm not sure what to rename it to.

Comment (david415) [reply internal/provider/kaetzchen/cbor_plugins.go]: what is a static method in golang?

File: cborplugin/client.go (hunk 4 above)
Comment (david415) [reply cborplugin/client.go]: i was thinking to not expose dial publicly because it is only useful in this context of dialing the unix domain socket AFTER execing the plugin.

File: internal/provider/kaetzchen/cbor_plugins.go (hunk 5 above)
Comment (willscott) [reply internal/provider/kaetzchen/cbor_plugins.go]: whoops, i misread- this is already static. the use of `glue` as grab-bag of context was the main issue

File: cborplugin/client.go (hunk 3 above)
Comment (david415) [reply cborplugin/client.go]: agreed

File: cborplugin/client.go (hunk 2 above)
Comment (david415) [reply cborplugin/client.go]: should it? i don't have a strong opinion either way... although the internal kaetzchen plugins take individual args right now:
https://github.com/katzenpost/server/blob/master/internal/provider/kaetzchen/kaetzchen.go#L76
as do the grpc plugins.

File: cborplugin/client.go (hunk 1 above)
Comment (david415) [reply cborplugin/client.go]: I agree. How about I rename it to PluginService?

File: cborplugin/client.go (hunk 2 above)
Comment (david415) [reply cborplugin/client.go]: i'd like for it to take the individual args here so that it matches the other plugin interfaces; the internal plugins and the grpc plugins.

File: cborplugin/client.go (hunk 6 above)
Comment (david415) [reply cborplugin/client.go]: It's an optional feature. Your plugin may return nil here and that's OK. Perhaps most plugins will not need this feature. It was useful for the "currency" plugin before most crypto currencies all use the same Bitcoin HTTP RPC; i wanted to reuse the same code but be able to publish which crypto currency is supported for each instance of this service. The ticker, BTC, ZEC is published in the MixDescriptor use this "Parameters" method.

File: internal/provider/kaetzchen/cbor_plugins.go (hunk 5 above)
Comment (david415) [reply internal/provider/kaetzchen/cbor_plugins.go]: yeah the glue usage is not good.

File: cborplugin/client.go (hunk 2 above)
Comment (nogoegst) [reply cborplugin/client.go]: yes, agree with @willscott here. in future I bet `Request` may be extended and still be passed to `OnRequest` then. otherwise you need to write another function to accept these new parameters.

File: cborplugin/client.go [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+
+// Response is the response received after sending a Request to the plugin.
+type Response struct {
+	Payload []byte
+}
+
+// Parameters is an optional mapping that plugins can publish, these get
+// advertised to clients in the MixDescriptor.
+type Parameters struct {
+	Map map[string]string
```
Comment (nogoegst) [inline cborplugin/client.go]: why not to just `type Parameters map[string]string`?

File: cborplugin/client.go [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+// Client acts as a client interacting with one or more plugins.
+type Client struct {
+	worker.Worker
+
+	log *logging.Logger
+
+	httpClient *http.Client
+	cmd        *exec.Cmd
+	socketPath string
+	params     map[string]string
```
Comment (nogoegst) [inline cborplugin/client.go]: And then here to have `Parameters` type from above.

File: cborplugin/client.go [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+	<-c.HaltCh()
+	c.cmd.Process.Signal(syscall.SIGHUP)
+	err := c.cmd.Wait()
+	if err != nil {
+		c.log.Errorf("CBOR plugin worker, command exec error: %s\n", err)
+	}
+}
+
+func (c *Client) dial(socketPath string) {
+	c.httpClient = &http.Client{
```
Comment (nogoegst) [inline cborplugin/client.go]: it is not clear from the function name that it just going to initialize http.Client with a dial function.

File: cborplugin/client.go [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+		Transport: &http.Transport{
+			DialContext: func(_ context.Context, _, _ string) (net.Conn, error) {
```
Comment (nogoegst) [inline cborplugin/client.go]: does it really make sense to ignore the context? AFAIK unix sockets are dialable with DialContext.

File: cborplugin/client.go [hunk 11]
```diff
@@ -0,0 +1,205 @@
...
+		return err
+	}
+	err = c.cmd.Start()
+	if err != nil {
+		return err
+	}
+
+	// read and decode plugin stdout
+	line := make([]byte, 255)
+	n, err := stdout.Read(line)
```
Comment (nogoegst) [inline cborplugin/client.go]: I expect stdout to be wrapped into ` bufio.Scanner` to read lines.

File: cborplugin/client.go [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+	if err != nil {
+		return err
+	}
+	socketPath := strings.TrimSuffix(string(line[:n]), "\n")
+	c.socketPath = socketPath
+	c.log.Debugf("plugin socket path:'%s'\n", socketPath)
+
+	c.dial(c.socketPath)
+	c.log.Debug("requesting plugin Parameters for Mix Descriptor publication...")
+	err = c.getParams()
```
Comment (nogoegst) [inline cborplugin/client.go]: either `getParams` should be named differently or return parameters (`params, err := c.getParams()`.

File: cborplugin/client.go [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+	if err != nil {
+		panic(err)
+	}
+	return nil
+}
+
+func (c *Client) getParams() error {
+	rawResponse, err := c.httpClient.Post("http://unix/parameters", "application/octet-stream", bytes.NewBuffer([]byte{}))
```
Comment (nogoegst) [inline cborplugin/client.go]: if you want to send no body, there is `net/http.NoBody` for this.

File: cborplugin/client.go [hunk 14]
```diff
@@ -0,0 +1,205 @@
...
+		ID:      id,
+		Payload: payload,
+		HasSURB: hasSURB,
+	}
+	var serialized []byte
+	enc := codec.NewEncoderBytes(&serialized, new(codec.CborHandle))
+	if err := enc.Encode(request); err != nil {
+		return nil, err
+	}
+	rawResponse, err := c.httpClient.Post("http://unix/request", "application/octet-stream", bytes.NewBuffer(serialized))
```
Comment (nogoegst) [inline cborplugin/client.go]: `bytes.NewReader` seems to be a better fit here.

File: cborplugin/client.go [hunk 15]
```diff
@@ -0,0 +1,205 @@
...
+	if err != nil {
+		return nil, err
+	}
+	response := Response{
+		Payload: []byte{},
```
Comment (nogoegst) [inline cborplugin/client.go]: does it need to be allocated?

File: cborplugin/client.go [hunk 16]
```diff
@@ -0,0 +1,205 @@
...
+	if err != nil {
+		return nil, err
+	}
+	return response.Payload, nil
+}
+
+// Parameters are used in Mix Descriptor publication to give
+// service clients more information about the service. Not
+// plugins will need to use this feature.
+func (c *Client) Parameters() map[string]string {
```
Comment (nogoegst) [inline cborplugin/client.go]: again, it's better to have `Parameters` type here and not opaque `map[string]string`.

File: config/config.go [hunk 17]
```diff
@@ -508,7 +512,57 @@ type PluginKaetzchen struct {
...
+
+	// Endpoint is the provider side endpoint that the agent will accept
+	// requests at.  While not required by the spec, this server only
+	// supports Endpoints that are lower-case local-parts of an e-mail
+	// address.
+	Endpoint string
+
+	// Config is the extra per agent arguments to be passed to the agent's
+	// initialization routine.
+	Config map[string]interface{}
```
Comment (nogoegst) [inline config/config.go]: maybe a new type here as well instead of map.

File: internal/provider/kaetzchen/cbor_plugins.go [hunk 18]
```diff
@@ -0,0 +1,258 @@
...
+// CBORPluginWorker is similar to Kaetzchen worker but uses
+// CBOR over HTTP over UNIX domain socket to talk to plugins.
+type CBORPluginWorker struct {
+	sync.Mutex
+	worker.Worker
+
+	glue glue.Glue
+	log  *logging.Logger
+
+	pluginChan map[[sConstants.RecipientIDLength]byte]*channels.InfiniteChannel
```
Comment (nogoegst) [inline internal/provider/kaetzchen/cbor_plugins.go]: this looks extra opaque to me.

File: internal/provider/kaetzchen/cbor_plugins.go [hunk 19]
```diff
@@ -0,0 +1,258 @@
...
+	clients    []*cborplugin.Client
+	forPKI     map[string]map[string]interface{}
```
Comment (nogoegst) [inline internal/provider/kaetzchen/cbor_plugins.go]: this to as well. I cannot see what are these nested key semantically. it may be specified as couple of new types which are basically strings.

File: internal/provider/kaetzchen/cbor_plugins.go [hunk 20]
```diff
@@ -0,0 +1,258 @@
...
+	if !ok {
+		k.log.Debugf("Failed to find handler. Dropping Kaetzchen request: %v", pkt.ID)
+		return
+	}
+	handlerCh.In() <- pkt
+}
+
+func (k *CBORPluginWorker) worker(recipient [sConstants.RecipientIDLength]byte, pluginClient cborplugin.ServicePlugin) {
+	// Kaetzchen delay is our max dwell time.
+	maxDwell := time.Duration(k.glue.Config().Debug.KaetzchenDelay) * time.Millisecond
```
Comment (nogoegst) [inline internal/provider/kaetzchen/cbor_plugins.go]: I just noticed that `KaetzchenDelay` is in milliseconds and `int`. why is that? why not just `time.Duration` which is just int64 of nanoseconds?

File: internal/provider/kaetzchen/cbor_plugins.go [hunk 21]
```diff
@@ -0,0 +1,258 @@
...
+	ct, surb, err := packet.ParseForwardPacket(pkt)
+	if err != nil {
+		k.log.Debugf("Dropping Kaetzchen request: %v (%v)", pkt.ID, err)
+		return
+	}
+
+	var resp []byte
+	respStr, err := pluginClient.OnRequest(pkt.ID, ct, surb != nil)
+	switch {
+	case err == nil:
```
Comment (nogoegst) [inline internal/provider/kaetzchen/cbor_plugins.go]: I'd rather make `switch err {` here.

File: internal/provider/kaetzchen/cbor_plugins.go [hunk 22]
```diff
@@ -0,0 +1,258 @@
...
+		k.log.Debugf("Processed Kaetzchen request: %v (No response)", pkt.ID)
+		return
+	default:
+		k.log.Debugf("Failed to handle Kaetzchen request: %v (%v), response: %s", pkt.ID, err, respStr)
+		return
+	}
+	resp = []byte(respStr)
+
+	// Iff there is a SURB, generate a SURB-Reply and schedule.
+	if surb != nil {
```
Comment (nogoegst) [inline internal/provider/kaetzchen/cbor_plugins.go]: it seems that early return should be here as a big block of code is behind just one condition.

File: internal/provider/kaetzchen/cbor_plugins.go [hunk 23]
```diff
@@ -0,0 +1,258 @@
...
+			}
+
+			// Accumulate a list of all clients to facilitate clean shutdown.
+			kaetzchenWorker.clients = append(kaetzchenWorker.clients, pluginClient)
+
+			// Start the worker.
+			worker := func() {
+				kaetzchenWorker.worker(endpoint, pluginClient)
+			}
+			kaetzchenWorker.Go(worker)
```
Comment (nogoegst) [inline internal/provider/kaetzchen/cbor_plugins.go]: I think is simpler to just do
```
kaetzchenWorker.Go(func() {
      kaetzchenWorker.worker(endpoint, pluginClient)
})
```

File: cborplugin/client.go (hunk 4 above)
Comment (mixmasala) [reply cborplugin/client.go]: I think the method should be renamed, it initializes the http client but does not dial.

File: internal/provider/kaetzchen/cbor_plugins.go (hunk 20 above)
Comment (mixmasala) [reply internal/provider/kaetzchen/cbor_plugins.go]: probably to make the configuration file more human readable

File: cborplugin/client.go (hunk 2 above)
Comment (david415) [reply cborplugin/client.go]: OK. I will change it to take the struct.

File: cborplugin/client.go (hunk 7 above)
Comment (david415) [reply cborplugin/client.go]: OK. Sounds good thanks.

File: cborplugin/client.go (hunk 8 above)
Comment (david415) [reply cborplugin/client.go]: ok yeah that sounds good too

File: cborplugin/client.go (hunk 9 above)
Comment (david415) [reply cborplugin/client.go]: ok i'll fix the name. i will call it, setupHttpClient.

File: cborplugin/client.go (hunk 10 above)
Comment (david415) [reply cborplugin/client.go]: hmm ok i will make it use DialContext

File: cborplugin/client.go (hunk 11 above)
Comment (david415) [reply cborplugin/client.go]: ok... i will use Scanner..

File: cborplugin/client.go (hunk 13 above)
Comment (david415) [reply cborplugin/client.go]: ok sounds much better

File: cborplugin/client.go (hunk 14 above)
Comment (david415) [reply cborplugin/client.go]: ok sounds good

File: cborplugin/client.go (hunk 15 above)
Comment (david415) [reply cborplugin/client.go]: not sure. i will try without it...

File: internal/provider/kaetzchen/cbor_plugins.go (hunk 23 above)
Comment (david415) [reply internal/provider/kaetzchen/cbor_plugins.go]: agreed

File: internal/provider/kaetzchen/cbor_plugins.go (hunk 22 above)
Comment (david415) [reply internal/provider/kaetzchen/cbor_plugins.go]: ok i put an early return there

File: internal/provider/kaetzchen/cbor_plugins.go (hunk 19 above)
Comment (david415) [reply internal/provider/kaetzchen/cbor_plugins.go]: ok i fixed this too

File: internal/provider/kaetzchen/cbor_plugins.go (hunk 18 above)
Comment (david415) [reply internal/provider/kaetzchen/cbor_plugins.go]: fixed

File: cborplugin/client.go (hunk 16 above)
Comment (david415) [reply cborplugin/client.go]: ok

**Comment by david415** [comment]
ok there. i made a bunch of corrections. i have still not managed to make a compatible rust plugin just yet. i will try again.

File: cborplugin/client.go [hunk 24]
```diff
@@ -131,75 +140,68 @@ func (c *Client) launch(command string, args []string) error {
...
 	if err != nil {
 		panic(err)
 	}
+	c.log.Debug("finished launching plugin.")
 	return nil
 }
 
-func (c *Client) getParams() error {
-	rawResponse, err := c.httpClient.Post("http://unix/parameters", "application/octet-stream", bytes.NewBuffer([]byte{}))
+func (c *Client) decodeParams() error {
```
Comment (nogoegst) [inline cborplugin/client.go]: what I expect from a func called `decodeSomething` is to decode something I pass there. this one issues an http request and decodes parameters from there.
I think there should be another func to fetch parameters and then another one to decode and return them along with the error. it's the caller responsibility to set them whenever.

File: cborplugin/client.go [hunk 25]
```diff
@@ -131,75 +140,68 @@ func (c *Client) launch(command string, args []string) error {
...
+	rawResponse, err := c.httpClient.Post("http://unix/parameters", "application/octet-stream", &http.NoBody)
```
Comment (nogoegst) [inline cborplugin/client.go]: I don't think it's necessary to pass NoBody via pointer.

File: cborplugin/client.go [hunk 26]
```diff
@@ -131,75 +140,68 @@ func (c *Client) launch(command string, args []string) error {
...
 	if err != nil {
+		c.log.Debugf("post failure: %s", err)
 		return err
 	}
-	response := Parameters{}
-	err = codec.NewDecoder(rawResponse.Body, new(codec.CborHandle)).Decode(&response)
+	responseParams := new(Parameters)
+	err = codec.NewDecoder(rawResponse.Body, new(codec.CborHandle)).Decode(&responseParams)
```
Comment (nogoegst) [inline cborplugin/client.go]: `new()` already returns a pointer.
I'd rather do it this way:
```golang
responseParams := &Parameters{}
err = codec.NewDecoder(rawResponse.Body, new(codec.CborHandle)).Decode(responseParams)
```

File: internal/provider/kaetzchen/cbor_plugins.go [hunk 27]
```diff
@@ -35,6 +35,21 @@ import (
 	"gopkg.in/op/go-logging.v1"
 )
 
+// PluginChan maps from Recipient ID to channel.
+type PluginChan = map[[sConstants.RecipientIDLength]byte]*channels.InfiniteChannel
```
Comment (nogoegst) [inline internal/provider/kaetzchen/cbor_plugins.go]: maybe `PluginChans`?

File: internal/provider/kaetzchen/cbor_plugins.go [hunk 28]
```diff
@@ -108,10 +123,15 @@ func (k *CBORPluginWorker) processKaetzchen(pkt *packet.Packet, pluginClient cbo
 	}
 
 	var resp []byte
-	respStr, err := pluginClient.OnRequest(pkt.ID, ct, surb != nil)
-	switch {
-	case err == nil:
-	case err == ErrNoResponse:
+	req := cborplugin.Request{
```
Comment (nogoegst) [inline internal/provider/kaetzchen/cbor_plugins.go]: I'd make it `&cborplugin.Request` here for consistency as well. otherwise somewhere it's a pointer and somewhere it's a value.

File: internal/provider/kaetzchen/cbor_plugins.go [hunk 29]
```diff
@@ -232,9 +254,11 @@ func NewCBORPluginWorker(glue glue.Glue) (*CBORPluginWorker, error) {
 				// just once we call the Parameters method on the plugin
 				// and use that info to populate our forPKI map which
 				// ends up populating the PKI document
-				p := pluginClient.Parameters()
-				for key, value := range p {
-					params[key] = value
+				p := pluginClient.GetParameters()
+				if p != nil {
+					for key, value := range *p {
```
Comment (nogoegst) [inline internal/provider/kaetzchen/cbor_plugins.go]: hmm, I think as p is `Parameters` which is a `map` which is a reference type. 
just this `range *p` looks weird to me. I can't recall if `range` will work over `Parameters` though.

Comment (david415) [reply internal/provider/kaetzchen/cbor_plugins.go]: range works for maps, since Parameters is a type alias for map this should work... and does build without compiler warnings

File: cborplugin/client.go (hunk 26 above)
Comment (david415) [reply cborplugin/client.go]: That won't work anymore because Parameters is now a type alias to a map.

File: cborplugin/client.go (hunk 25 above)
Comment (david415) [reply cborplugin/client.go]: correct. it builds without the reference

File: cborplugin/client.go (hunk 26 above)
Comment (nogoegst) [reply cborplugin/client.go]: but then it should be `make`d, no?

File: internal/provider/kaetzchen/cbor_plugins.go [hunk 30]
```diff
@@ -185,11 +184,11 @@ func (k *CBORPluginWorker) launch(command string, args []string) (*cborplugin.Cl
...
 
 	kaetzchenWorker := CBORPluginWorker{
-		glue:       glue,
-		log:        glue.LogBackend().GetLogger("CBOR plugin worker"),
-		pluginChan: make(map[[sConstants.RecipientIDLength]byte]*channels.InfiniteChannel),
-		clients:    make([]*cborplugin.Client, 0),
-		forPKI:     make(map[string]map[string]interface{}),
+		glue:        glue,
+		log:         glue.LogBackend().GetLogger("CBOR plugin worker"),
+		pluginChans: make(map[[sConstants.RecipientIDLength]byte]*channels.InfiniteChannel),
```
Comment (nogoegst) [inline internal/provider/kaetzchen/cbor_plugins.go]: `make` should also accept `PluginChans` type. probably same goes for `forPKI`.

File: cborplugin/client.go [hunk 31]
```diff
@@ -0,0 +1,201 @@
...
+	if err != nil {
+		c.log.Errorf("CBOR plugin worker, command exec error: %s\n", err)
+	}
+}
+
+func (c *Client) setupHTTPClient(socketPath string) {
+	c.httpClient = &http.Client{
+		Transport: &http.Transport{
+			DialContext: func(context context.Context, _, _ string) (net.Conn, error) {
+				return new(net.Dialer).DialContext(context, "unix", socketPath)
```
Comment (nogoegst) [inline cborplugin/client.go]: it's better not to shadow package names, so `ctx` should be here for the var name.

File: cborplugin/client.go [hunk 32]
```diff
@@ -0,0 +1,201 @@
...
+		},
+	}
+}
+
+func (c *Client) launch(command string, args []string) error {
+	// exec plugin
+	if args == nil {
+		c.cmd = exec.Command(command)
+	} else {
+		c.cmd = exec.Command(command, args...)
```
Comment (nogoegst) [inline cborplugin/client.go]: I don't think one has to check for `nil` here actually. it should work just with `command, args...`.

File: cborplugin/client.go [hunk 33]
```diff
@@ -0,0 +1,201 @@
...
+	err = c.cmd.Start()
+	if err != nil {
+		c.log.Debugf("failed to exec: %s", err)
+		return err
+	}
+
+	// read and decode plugin stdout
+	stdoutScanner := bufio.NewScanner(stdout)
+	stdoutScanner.Scan()
+	c.socketPath = stdoutScanner.Text()
```
Comment (nogoegst) [inline cborplugin/client.go]: it's not a big deal, but I would leave reading stdout of something as a separate function. like `readSocketPath()` or similar.

File: cborplugin/client.go [hunk 34]
```diff
@@ -0,0 +1,201 @@
...
+	}
+	c.params = &responseParams
+	c.log.Debug("finished launching plugin.")
+	return nil
+}
+
+// OnRequest send a query request to plugin using CBOR + HTTP over Unix domain socket.
+func (c *Client) OnRequest(request *Request) ([]byte, error) {
+	var serialized []byte
+	enc := codec.NewEncoderBytes(&serialized, new(codec.CborHandle))
```
Comment (nogoegst) [inline cborplugin/client.go]: wait, `serialized` is `nil` here. but then comes the reference to it.
it will make more sense to actually allocate some slice like `[]byte{}`.

File: internal/provider/kaetzchen/cbor_plugins.go [hunk 35]
```diff
@@ -0,0 +1,280 @@
...
+		return
+	}
+	ch := handlerCh.Out()
+
+	for {
+		var pkt *packet.Packet
+		select {
+		case <-k.HaltCh():
+			k.log.Debugf("Terminating gracefully.")
+			k.killAllClients()
```
Comment (nogoegst) [inline internal/provider/kaetzchen/cbor_plugins.go]: oh my :)  can this function be renamed to something less violent? `stopAllClients`, `shutdownAllClients` etc?

File: internal/provider/kaetzchen/cbor_plugins.go [hunk 36]
```diff
@@ -0,0 +1,280 @@
...
+				pkt.Dispose()
+				continue
+			}
+		}
+
+		k.processKaetzchen(pkt, pluginClient)
+	}
+}
+
+func (k *CBORPluginWorker) killAllClients() {
```
Comment (nogoegst) [inline internal/provider/kaetzchen/cbor_plugins.go]: yeah, even `haltAllClients` will make sense.

File: internal/provider/kaetzchen/cbor_plugins.go [hunk 37]
```diff
@@ -0,0 +1,280 @@
...
+	case ErrNoResponse:
+		k.log.Debugf("Processed Kaetzchen request: %v (No response)", pkt.ID)
+		return
+	default:
+		k.log.Debugf("Failed to handle Kaetzchen request: %v (%v), response: %s", pkt.ID, err, respStr)
+		return
+	}
+	resp = []byte(respStr)
+
+	if resp == nil {
```
Comment (nogoegst) [inline internal/provider/kaetzchen/cbor_plugins.go]: it will never be `nil`. there should be empty string comparison above or `len() == 0` check here.

File: internal/provider/kaetzchen/cbor_plugins.go [hunk 38]
```diff
@@ -0,0 +1,280 @@
...
+
+		k.log.Debugf("Handing off newly generated SURB-Reply: %v (Src:%v)", respPkt.ID, pkt.ID)
+		k.glue.Scheduler().OnPacket(respPkt)
+		return
+	}
+	k.log.Debugf("No SURB provided: %v", pkt.ID)
+}
+
+// KaetzchenForPKI returns the plugins Parameters map for publication in the PKI doc.
+func (k *CBORPluginWorker) KaetzchenForPKI() map[string]map[string]interface{} {
```
Comment (nogoegst) [inline internal/provider/kaetzchen/cbor_plugins.go]: i guess this should return `ServiceMap` now.h

**Comment by david415** [comment]
there, i made another round of corrections. is this good enough to merge? soon later we shall replace this with a CBOR + websocket plugin system, because we need bidirectional messaging for our CRDT pub sub protocol.
